i18n/FlippableHandler: Automatically adapt the height of the element to fit it’s content

mraveux opened the pull request:
Because of i18n some translations caused the content to overflow on the bottom, and that broke design. With this commit now it automatically adapt the height of the `#rotation-container` element so that it fit its content and do not break design anymore.


_Btw, in the `FlippableHandler` you used a lot `window.setTimeout(..., 0)`.
Is there a reason you used this instead of `requestAnimationFrame` ?_

mraveux committed:
i18n/FlippableHandler: Automatically adapt the height of the element to fit it’s content

Because of i18n some translations caused the content to overflow on the bottom, and that broke design. With this commit, now it automatically adapt the height of the #rotation-container element so that it fit its content and do not break design anymore.

sisou commented on src/components/FlippableHandler.js:
```diff
@@ -32,21 +33,60 @@ class FlippableHandler {
+                const $visiblePageHeight = $pages.reduce(
+                    (acc, $el) => ($el.offsetParent && $el.clientHeight > acc ? $el.clientHeight : acc),
+                    0,
+                );
```
You can find the biggest height with
```javascript
const heights = $pages.map(($el) => $el.offsetParent ? $el.clientHeight : 0);

Math.max.apply(Math, heights);
// or
Math.max(...heights);
```
which is better/simpler than `.reduce()`.

sisou commented on src/components/FlippableHandler.js:
```diff
@@ -32,21 +33,60 @@ class FlippableHandler {
+                    0,
+                );
+                if ($visiblePageHeight > 0) {
+                    $rotationContainer.style.height = `${$visiblePageHeight}px`;
```
Would it make sense to revert it to its default (or `auto`), when no height was found?

danimoh commented on src/components/FlippableHandler.js:
```diff
@@ -32,21 +33,60 @@ class FlippableHandler {
                         }, 0);
+                    } else {
+                        window.setTimeout(() => {
+                            $rotationContainer.classList.add('disable-transition');
```
Are the timeouts and setting / unsetting `disable-transition` necessary here?

mraveux replied:
The first timeout is not needed anymore, you're right.
But the second timeout as well as setting / unsetting `disable-transition` is needed to avoid animating the height change at the loading of the page.

mraveux committed:
FlippableHandler: PR #405  - Requested changes

Requested changes by Sisou
https://github.com/nimiq/keyguard/pull/405#pullrequestreview-458219910

- Now using `.map` & `Math.max` instead of `.reduce`
- Now reverting the height back to default if the found height is not above 0

danimoh commented on src/components/FlippableHandler.js:
```diff
@@ -32,21 +33,60 @@ class FlippableHandler {
+    /**
+     * Update the height of the #rotation-container element to match its content.
+     * The default behavior is to look for every visible `.page` element,
+     * and, if there is multiple, take the height of the higher.
```
`if there are multiple`

mraveux replied:
😅

danimoh commented on src/components/FlippableHandler.js:
```diff
@@ -32,21 +33,56 @@ class FlippableHandler {
+            /** @type {Array<HTMLElement>} */
+            const $pages = Array.from($rotationContainer.querySelectorAll('.page'));
+            if ($pages && $pages.length > 0) {
+                const heights = $pages.map($el => ($el.offsetParent ? $el.clientHeight : 0));
```
The check here isn't necessary as `clientHeight` is automatically 0 if the element is not attached to the DOM.

mraveux committed:
FlippableHandler: PR #405  - Requested changes

Requested changes by danimoh
https://github.com/nimiq/keyguard/pull/405#pullrequestreview-458391217

- remove an unecessary `setTimeout`
- remove the check for `offsetParent` since `clientHeight` equals `0` if the element is not attached to the DOM / not visible
- minor comment correction

sisou reviewed (approved):
Superb! :+1:
